marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.smart-mastering/survivorship/merging/base.xqy

```diff
@@ -317,12 +319,18 @@ declare function merge-impl:generate-provenance-details(
 ) as item()*
 {
   map:new((
-    for $source in fn:distinct-values(($final-properties ! map:get(., "sources")/documentUri))
+    let $properties-by-doc-uri := map:map()
```

Comment:
Group properties by document URIs once in the beginning to avoid XPath that would repeatedly check all the properties with an XPath predicate.